File: legacy/cb/dtos/node.py
from __future__ import annotations
"""Node / Agent DTOs."""

import dataclasses
from cb.dtos.base import BaseDTO


@dataclasses.dataclass
class NodeDTO(BaseDTO):
    name: str = ""
    display_name: str = ""
    offline: bool = False
    num_executors: int = 1
    labels: str = ""
    description: str = ""

    @classmethod
    def from_dict(cls, data: dict) -> "NodeDTO":
        return cls(
            name=data.get("displayName", data.get("name", "")),
            display_name=data.get("displayName", ""),
            offline=data.get("offline", False),
            num_executors=data.get("numExecutors", 1),
            labels=data.get("assignedLabels", [{}])[0].get("name", "") if data.get("assignedLabels") else "",
            description=data.get("description", ""),
        )


@dataclasses.dataclass
class NodeDetailDTO(NodeDTO):
    launcher_type: str = ""   # jnlp / ssh / command
    remote_dir: str = ""
    config_xml: str = ""      # raw XML, used for copy-node

    @classmethod
    def from_dict(cls, data: dict) -> "NodeDetailDTO":
        base = NodeDTO.from_dict(data)
        launcher = data.get("launcher", {})
        launcher_class = launcher.get("_class", "") if isinstance(launcher, dict) else ""
        if "JNLP" in launcher_class or "Inbound" in launcher_class:
            ltype = "jnlp"
        elif "SSH" in launcher_class:
            ltype = "ssh"
        else:
            ltype = launcher_class.split(".")[-1].lower()

        return cls(
            name=base.name,
            display_name=base.display_name,
            offline=base.offline,
            num_executors=base.num_executors,
            labels=base.labels,
            description=base.description,
            launcher_type=ltype,
            remote_dir=data.get("remoteFS", ""),
        )

File: src/core/dtos/node.ts
/**
 * Node / Agent DTOs.
 * Mirrors legacy/cb/dtos/node.py
 */

import { str, num, bool, nested } from "./base.js";

/** Shape returned by the agents list endpoint (`/computer/api/json` `computer[]` entries). */
export interface NodeDTO {
  name: string;
  displayName: string;
  offline: boolean;
  numExecutors: number;
  labels: string;
  description: string;
}

/**
 * Extended node shape from `/computer/<name>/api/json` (single-agent detail).
 * Adds `launcherType`, `remoteDir`, and `configXml` on top of NodeDTO.
 */
export interface NodeDetailDTO extends NodeDTO {
  launcherType: string;
  remoteDir: string;
  configXml: string;
}

/**
 * Maps a raw agent list entry to NodeDTO.
 * `assignedLabels[0].name` → `labels`; `displayName` falls back to `name` when blank.
 */
export function nodeFromDict(data: Record<string, unknown>): NodeDTO {
  const assignedLabels = data["assignedLabels"];
  let labels = "";
  if (Array.isArray(assignedLabels) && assignedLabels.length > 0) {
    labels = str(nested(assignedLabels[0] as Record<string, unknown>, "name"));
  }

  const displayName = str(data["displayName"]);
  return {
    name: displayName || str(data["name"]),
    displayName,
    offline: bool(data["offline"], false),
    numExecutors: num(data["numExecutors"], 1),
    labels,
    description: str(data["description"]),
  };
}

function classToLauncherType(launcherClass: string): string {
  if (launcherClass.includes("JNLP") || launcherClass.includes("Inbound"))
    return "jnlp";
  if (launcherClass.includes("SSH"))
    return "ssh";
  if (!launcherClass) return "";
  return launcherClass.split(".").at(-1)!.toLowerCase();
}

/**
 * Maps a raw single-agent detail response to NodeDetailDTO.
 * `launcher._class` → normalized `launcherType` (jnlp/ssh/…); `remoteFS` → `remoteDir`.
 */
export function nodeDetailFromDict(data: Record<string, unknown>): NodeDetailDTO {
  const base = nodeFromDict(data);
  const launcher = data["launcher"];
  const launcherClass =
    launcher !== null && launcher !== undefined && typeof launcher === "object"
      ? str((launcher as Record<string, unknown>)["_class"])
      : "";

  return {
    ...base,
    launcherType: classToLauncherType(launcherClass),
    remoteDir: str(data["remoteFS"]),
    configXml: str(data["configXml"]),
  };
}
